File: BKJN_CT/SmallLighter/2667.js
const filePath =
  process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
const input = require('fs')
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n');

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

// 연결 집합의 개수와 그 집합을 구성하는 요소의 개수를 구하는 문제

const N = +input.shift();

let graph = input.map((row) => row.split('').map(Number));

let visited = Array.from(Array(N), () => Array(N).fill(false));

let cntHome = 0;
let cntComplex = 0;
const answer = [];

const dfs = (x, y) => {
  if (graph[x][y] === 0 || visited[x][y]) return;

  visited[x][y] = true;
  cntHome++;

  for (let i = 0; i < 4; i++) {
    const nx = x + dx[i];
    const ny = y + dy[i];
    if (nx >= 0 && nx < N && ny >= 0 && ny < N) dfs(nx, ny);
  }
};

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    if (graph[i][j] === 1 && !visited[i][j]) {
      dfs(i, j);
      cntComplex++;
      answer.push(cntHome);
      cntHome = 0;
    }
  }
}

console.log(cntComplex + '\n' + answer.sort((a, b) => a - b).join('\n'));

File: BKJN_CT/SmallLighter/2667.py
import sys
from collections import deque

sys.stdin = open('./BKJN_CT/SmallLighter/input.txt', 'r')

input = sys.stdin.readline

N = int(input())

graph = []

for _ in range(N):
    graph.append(list(map(int, input().rstrip())))

dx = [-1,1,0,0]
dy = [0,0,1,-1]

cntComplex = 0
homeRes = []

def BFS(x,y):
    q = deque([[x,y]])
    cntHome = 1

    while(q):
        x,y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < N and 0 <= ny < N and graph[nx][ny] == 1:
                q.append([nx,ny])
                graph[nx][ny] = 0
                cntHome += 1
    homeRes.append(cntHome)
    

for i in range(N):
    for j in range(N):
        if graph[i][j] == 1:
            graph[i][j] = 0
            BFS(i,j)
            cntComplex += 1


homeRes.sort()
print(cntComplex)
print('\n'.join(str(s) for s in homeRes))